File: sdks/ruby/lib/autonoma/refs.rb
# frozen_string_literal: true

require "openssl"
require "base64"
require "json"
require "bigdecimal"
require "date"
require "time"

module Autonoma
  # Sign / verify the teardown token: a JWT-like token (header.payload.signature)
  # over HMAC-SHA256 - not a full JWT library, to avoid dependencies.
  #
  # The payload is { "refs" => <arbitrary JSON>, "testRunId" => <string>,
  # "environment" => <scenario name> }. refs is whatever a scenario's up
  # returned; it is signed at up and handed back to the scenario's down verbatim.
  # environment carries the scenario name so down can route to the right
  # teardown (named "environment" for wire/back-compat reasons).
  module Refs
    def self.sign_refs(payload, secret)
      header = base64url_encode(JSON.generate({ alg: "HS256", typ: "REFS" }))
      body = base64url_encode(JSON.generate(make_json_safe(payload)))
      signature = hmac_sign("#{header}.#{body}", secret)
      "#{header}.#{body}.#{signature}"
    end

    # Verify and decode a teardown token. Returns the payload hash or raises.
    def self.verify_refs(token, secret)
      parts = token.split(".")
      raise "malformed token" unless parts.length == 3

      header, body, signature = parts
      expected = hmac_sign("#{header}.#{body}", secret)

      raise "signature mismatch" unless Autonoma.secure_compare(expected, signature)

      JSON.parse(base64url_decode(body))
    end

    def self.base64url_encode(data)
      data = data.encode("UTF-8") if data.is_a?(String)
      Base64.urlsafe_encode64(data, padding: false)
    end

    def self.base64url_decode(data)
      padding = 4 - (data.length % 4)
      data += "=" * padding if padding != 4
      Base64.urlsafe_decode64(data)
    end

    def self.hmac_sign(data, secret)
      sig = OpenSSL::HMAC.digest("SHA256", secret, data)
      Base64.urlsafe_encode64(sig, padding: false)
    end

    # Recursively convert non-JSON-safe types (Time, DateTime, BigDecimal, ...)
    # to strings so JSON.generate does not raise. The Rails adapter reuses this
    # to serialize response bodies that may carry such values.
    def self.make_json_safe(obj)
      case obj
      when Hash
        obj.transform_values { |v| make_json_safe(v) }
      when Array
        obj.map { |v| make_json_safe(v) }
      when Time, DateTime
        obj.iso8601(3)
      when Date
        obj.iso8601
      when BigDecimal
        obj.to_s("F")
      when Symbol
        obj.to_s
      else
        obj
      end
    end

    private_class_method :base64url_encode, :base64url_decode, :hmac_sign
  end
end

File: sdks/python/src/autonoma/refs.py
"""JWT-like teardown token: header.payload.signature using HMAC-SHA256."""

import base64
import hashlib
import hmac
import json

from autonoma.serializer import default_serializer as _default_serializer


def sign_refs(payload: dict, secret: str) -> str:
    """Sign a refs payload into a 3-part token string.

    In v2 the payload carries ``refs`` (whatever the scenario returned),
    ``testRunId``, and ``environment`` (the scenario name, so ``down`` can
    route to the right teardown).
    """
    header = _base64url_encode(json.dumps({"alg": "HS256", "typ": "REFS"}, separators=(",", ":")).encode())
    body = _base64url_encode(json.dumps(payload, separators=(",", ":"), default=_default_serializer).encode())
    signature = _hmac_sign(f"{header}.{body}", secret)
    return f"{header}.{body}.{signature}"


def verify_refs(token: str, secret: str) -> dict:
    """Verify and decode a teardown token. Returns the payload dict or raises."""
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("malformed token")

    header, body, signature = parts
    expected = _hmac_sign(f"{header}.{body}", secret)

    if not hmac.compare_digest(expected, signature):
        raise ValueError("signature mismatch")

    return json.loads(_base64url_decode(body))


def _base64url_encode(data: bytes) -> str:
    """Base64url encode without padding."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _base64url_decode(data: str) -> bytes:
    """Base64url decode, adding padding back."""
    padding = 4 - len(data) % 4
    if padding != 4:
        data += "=" * padding
    return base64.urlsafe_b64decode(data)


def _hmac_sign(data: str, secret: str) -> str:
    """HMAC-SHA256 sign and return base64url-encoded result without padding."""
    sig = hmac.new(secret.encode(), data.encode(), hashlib.sha256).digest()
    return base64.urlsafe_b64encode(sig).rstrip(b"=").decode()
